src/upload.js

```diff
@@ -4,46 +4,115 @@ const FormData = require("form-data");
 const fs = require("fs");
 const p = require("path");
 
-const { defaultOptions, walkDirectory, uriSkynetPrefix } = require("./utils");
-const { SkynetClient } = require("./client");
+const { Upload } = require("tus-js-client");
+
+const { buildRequestHeaders, SkynetClient } = require("./client");
+const { defaultOptions, getFileMimeType, makeUrl, walkDirectory, uriSkynetPrefix } = require("./utils");
+
+/**
+ * The tus chunk size is (4MiB - encryptionOverhead) * dataPieces, set in skyd.
+ */
+const TUS_CHUNK_SIZE = (1 << 22) * 10;
+
+/**
+ * The retry delays, in ms. Data is stored in skyd for up to 20 minutes, so the
+ * total delays should not exceed that length of time.
+ */
+const DEFAULT_TUS_RETRY_DELAYS = [0, 5000, 15000, 60000, 300000, 600000];
 
 const defaultUploadOptions = {
   ...defaultOptions("/skynet/skyfile"),
+  endpointLargeUpload: "/skynet/tus",
+
   portalFileFieldname: "file",
   portalDirectoryFileFieldname: "files[]",
   customFilename: "",
   customDirname: "",
   dryRun: false,
+  largeFileSize: TUS_CHUNK_SIZE,
+  retryDelays: DEFAULT_TUS_RETRY_DELAYS,
 };
 
-SkynetClient.prototype.uploadFile = function (path, customOptions = {}) {
+SkynetClient.prototype.uploadFile = async function (path, customOptions = {}) {
   const opts = { ...defaultUploadOptions, ...this.customOptions, ...customOptions };
 
-  const formData = new FormData();
-  const filename = opts.customFilename ? opts.customFilename : "";
-  formData.append(opts.portalFileFieldname, fs.createReadStream(path), filename);
+  const sizeInBytes = fs.statSync(path).size;
+  const filename = opts.customFilename ? opts.customFilename : p.basename(path);
 
+  if (sizeInBytes < opts.largeFileSize) {
+    return await uploadSmallFile(this, path, filename, opts);
+  } else {
+    return await uploadLargeFile(this, path, filename, sizeInBytes, opts);
+  }
+};
+
+async function uploadSmallFile(client, path, filename, opts) {
   const params = {};
   if (opts.dryRun) params.dryrun = true;
 
+  const formData = new FormData();
+  formData.append(opts.portalFileFieldname, fs.createReadStream(path), filename);
```

Comment:
should we validate the `portalFileFieldname` is not empty?

Reply:
We don't really do any validation of inputs in this SDK. I don't think it's worth hardening this one like we did with `skynet-js` as this will be archived once we have an isomorphic SDK.

Reply:
not sure if that's a good enough reason to not do basic validation on inputs

Reply:
In this particular case though I guess it's fine to leave it as is.